### ISSUE TITLE
Automatic snapshot generator

### ISSUE THREAD

**Issue by plehegar** [w3c/modern-tooling#75]
Idea here is to have the ability to generate snapshots based on git commits.

**Comment by xfq** [comment]
Currently, we can already link to snapshots with [RawGit](https://rawgit.com/)/[HTMLPreview](https://htmlpreview.github.io/) (without the [WHATWG warning](https://github.com/tabatkins/bikeshed/blob/6f5fe05144647376f0e7a3006f9a181963f53c1d/bikeshed/boilerplate/warning-commit.include), tho), because they support using a specific tag or commit hash in the URL.

However, one issue with RawGit/HTMLPreview is, if the spec uses Bikeshed and auto-deploys the generated `index.html` to `gh-pages` with Travis CI, you need to find the corresponding commit in `gh-pages`, rather than using the commit in `master` directly.

-----

Our friends at WHATWG do this in build-time: https://github.com/whatwg/whatwg.org/blob/e8cda18f1f9168b525f9612ce4ca49d7a92f6430/resources.whatwg.org/build/deploy.sh#L93-L103

They also include a client-side [script](https://github.com/whatwg/whatwg.org/blob/master/resources.whatwg.org/commit-snapshot-shortcut-key.js) in the [boilerplate](https://github.com/tabatkins/bikeshed/blob/150b038a54cb926cf8e3d36fd8b108c226cb94b8/bikeshed/boilerplate/whatwg/header.include#L11), letting the spec reader go to the commit snapshot by pressing `y` (similar to [GitHub](https://help.github.com/articles/getting-permanent-links-to-files/#press-y-to-permalink-to-a-file-in-a-specific-commit)). (Bikeshed PR: https://github.com/tabatkins/bikeshed/pull/288)

**Comment by plehegar** [comment]
Here is what I think we'd need: provide a link to the commit snapshot, provide a history of snapshots, works with respec and bikeshed, easily deployable.

**Comment by xfq** [comment]
If we don't want to use RawGit/HTMLPreview, should we save the commit snapshots on GitHub (not in the default branch)? If so, we will have 1,000 separate files if there are 1,000 commits (and that doesn't even count the multiple-file spec case). [PR Preview](https://github.com/tobie/pr-preview) stores the specs on AWS, but AWS is not easy to access in China...

**Comment by xfq** [comment]
I just created a repo for this: https://github.com/xfq/commit-snapshot-tool

**Comment by plehegar** [comment]
if we need a cloud server, maybe we can use a labs.w3.org?

**Comment by tabatkins** [comment]
Note that [pr-preview](https://github.com/tobie/pr-preview) already exists and is *almost* this functionality; it works with Bikeshed and ReSpec and generates previews for any PR.

**Comment by xfq** [comment]
> Note that [pr-preview](https://github.com/tobie/pr-preview) already exists and is *almost* this functionality; it works with Bikeshed and ReSpec and generates previews for any PR.

Yep. It uses a [JSON file](https://github.com/tobie/pr-preview/blob/2bc932997ebe66aebccb0f089f11d0f9d1840281/README.md#configuration-file) to reflect the configuration of the spec, and uses the [Bikeshed CGI](https://github.com/tobie/pr-preview/blob/2bc932997ebe66aebccb0f089f11d0f9d1840281/lib/mixins/fetchable.js#L27) and [Spec Generator](https://github.com/tobie/pr-preview/blob/2bc932997ebe66aebccb0f089f11d0f9d1840281/lib/mixins/fetchable.js#L33) with [raw links on GitHub](https://github.com/tobie/pr-preview/blob/2bc932997ebe66aebccb0f089f11d0f9d1840281/lib/mixins/fetchable.js#L41) and [RawGit](https://github.com/tobie/pr-preview/blob/2bc932997ebe66aebccb0f089f11d0f9d1840281/lib/mixins/fetchable.js#L45) to generate the spec, and saves the generated spec on [AWS](https://github.com/tobie/pr-preview/blob/2bc932997ebe66aebccb0f089f11d0f9d1840281/lib/cache.js).

**Comment by xfq** [comment]
And drafts.csswg.org also has this functionality, e.g., https://drafts.csswg.org/history/css-values-4/

([Source](https://hg.csswg.org/dev/drafts/file/5ebfbd25a149/pages/HistoryPage.php))

**Comment by xfq** [comment]
Maybe we can reuse some code in the projects above, and deploy it to labs.w3.org.

**Comment by xfq** [comment]
@tobie Could you provide some hint on how we should build this tool (or even better, we can collaborate on extending pr-preview)?

I think we should monitor the [push](https://developer.github.com/v3/activity/events/types/#pushevent) event, detect the config of the repo (either by reusing `.pr-preview.json`, or by creating a new config file), build the spec using Bikeshed/ReSpec's web service, and save the generated snapshot on some server. We probably also need frontend code to display a history of snapshots of a given spec.

**Comment by tobie** [comment]
I agree you can probably reuse large chunks of pr-preview for this. And the overall architecture you describe looks good to me.

I'd be happy to help, but that would require setting up some consultancy agreement which I doubt W3C has funds for.

> PR Preview stores the specs on AWS, but AWS is not easy to access in China...

Oh! That's interesting. Is the filtering happening on the domain name or on the actual IP? If it's the former, adding custom domain names is easy. That's what we've done for WHATWG, e.g.: https://whatpr.org/html/3714/iframe-embed-object.html is hosted on AWS S3.

**Comment by xfq** [comment]
> > PR Preview stores the specs on AWS, but AWS is not easy to access in China...
> 
> Oh! That's interesting. Is the filtering happening on the domain name or on the actual IP? If it's the former, adding custom domain names is easy. That's what we've done for WHATWG, e.g.: https://whatpr.org/html/3714/iframe-embed-object.html is hosted on AWS S3.

It's a sophisticated system, consisting of DNS filtering, IP blocking, TCP ports blocking, and other blocking methods. Some (but not all) AWS servers are not accessible in some regions/ISPs in China, and the list of servers being blocked is constantly changing.

For example, in my ISP:

* The DNS seems to correctly resolve whatpr.org and pr-preview.s3.amazonaws.com
* I can successfully ping whatpr.org and pr-preview.s3.amazonaws.com (ICMP works correctly)
* I can't establish TCP connections to whatpr.org now (so that HTTP and SSH connections are blocked), but I can establish TCP connections to pr-preview.s3.amazonaws.com
